### PR TITLE
Add LlavaImageProcessor

### LINKED ISSUE #33175: Supporting Padding in llava processor
### Feature request

https://github.com/haotian-liu/LLaVA uses padding for pre-processing the images by default. Current transformers implementation does not support that.

### Motivation

Request per @NielsRogge at (https://huggingface.co/llava-hf/llava-1.5-7b-hf/discussions/26#66cf46a5a523b74b5f90fa72).

### Your contribution

I successfully reproduced logits after conversion if we add padding in the Transformers library.

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

Fixes #33175. It adds the option to pad an image before applying the same preprocessing as `CLIPImageProcessor` based on the [original implementation](https://github.com/haotian-liu/LLaVA/blob/c121f0432da27facab705978f83c4ada465e46fd/llava/mm_utils.py#L152). This allows people to match the logits with the original implementation.

For now I decided to set `do_pad` to `False` by default as otherwise it would be a breaking change.

To do:

- [x] update padding method to take numpy as input and produce numpy as output
- [x] add equivalence test
- [x] add tests

### CODE REVIEW

File: src/transformers/models/llava/image_processing_llava.py [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+        max_dim = max(height, width)
+        pad_height = max_dim - height
+        pad_width = max_dim - width
+
+        padding = (
+            (pad_height // 2, pad_height - pad_height // 2),
+            (pad_width // 2, pad_width - pad_width // 2),
+        )
+
+        return pad(
```
Comment (NielsRogge) [inline src/transformers/models/llava/image_processing_llava.py:216]: cc @amyeroberts I'd like to replicate [this function](https://github.com/haotian-liu/LLaVA/blob/c121f0432da27facab705978f83c4ada465e46fd/llava/mm_utils.py#L152) using the `pad` function of image_transforms.py. However, it seems padding the channel dimension with custom values isn't supported (they pad using background_color = (122, 116, 104) as seen [here](https://github.com/haotian-liu/LLaVA/blob/c121f0432da27facab705978f83c4ada465e46fd/llava/mm_utils.py#L171)).

Is this something we could add support for?

Comment (amyeroberts) [reply src/transformers/models/llava/image_processing_llava.py:216]: I'd just write a custom `pad` function in this module. If it turns out being a functionality which is nedded for a lot of other models, then we can think about adding into the more general image_transforms module

File: src/transformers/models/llava/image_processing_llava.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+            "input_data_format",
+        ]
+
+        # for backwards compatibility of KOSMOS-2
+        if "use_square_size" in kwargs and kwargs["use_square_size"]:
+            self.size = {"height": size["shortest_edge"], "width": size["shortest_edge"]}
+            # Let's remove `use_square_size` (as it is removed from #27690), so the future Kosmos-2 image processors
+            # won't have this attr. being saved. (otherwise, it will enter this if branch while there is no more
+            # `shortest_edge` key.
+            delattr(self, "use_square_size")
```
Comment (zucchini-nlp) [inline src/transformers/models/llava/image_processing_llava.py:155]: shouldn't be needed? Seems like it comes from `KosmosImageProcessor` which relies on Clip processor's code

**Review by zucchini-nlp — APPROVED** [review]
Thanks, looks good to me! Wondering if we need to add a small tip section in the docs, to make users aware of padding (in case they want to enable 100% match with original impl)?

File: src/transformers/models/llava/image_processing_llava.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+    ) -> PIL.Image.Image:
+        """
+        Preprocess an image or batch of images.
+
+        Args:
+            images (`ImageInput`):
+                Image to preprocess. Expects a single or batch of images with pixel values ranging from 0 to 255. If
+                passing in images with pixel values between 0 and 1, set `do_rescale=False`.
+            do_pad (`bool`, *optional*, defaults to `self.do_pad`):
+                Whether to pad the image to a square.
```
Comment (amyeroberts) [inline src/transformers/models/llava/image_processing_llava.py:287]: Should be clarified the output shape resulting from the padding

File: src/transformers/models/llava/image_processing_llava.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+        ]
+
+    def pad_to_square(
+        self,
+        image: np.array,
+        background_color: Union[int, Tuple[int, int, int]] = 0,
+        input_data_format: Optional[Union[str, ChannelDimension]] = None,
+    ) -> np.array:
+        """
+        Pads an image to a square.
```
Comment (amyeroberts) [inline src/transformers/models/llava/image_processing_llava.py:156]: Should be clarified the output shape resulting from the padding

File: src/transformers/models/llava/image_processing_llava.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+    import PIL
+
+
+class LlavaImageProcessor(BaseImageProcessor):
+    r"""
+    Constructs a LLaVa image processor.
+
+    Args:
+        do_pad (`bool`, *optional*, defaults to `False`):
+            Whether to pad the image to a square. Can be overridden by `do_pad` in the `preprocess` method.
```
Comment (amyeroberts) [inline src/transformers/models/llava/image_processing_llava.py:59]: Should be clarified the output shape resulting from the padding

File: tests/models/llava/test_image_processing_llava.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        image_processor = self.image_processing_class.from_dict(self.image_processor_dict)
+        self.assertEqual(image_processor.size, {"shortest_edge": 20})
+        self.assertEqual(image_processor.crop_size, {"height": 18, "width": 18})
+
+        image_processor = self.image_processing_class.from_dict(self.image_processor_dict, size=42, crop_size=84)
+        self.assertEqual(image_processor.size, {"shortest_edge": 42})
+        self.assertEqual(image_processor.crop_size, {"height": 84, "width": 84})
+
+    # Ignore copy
+    def test_padding(self):
```
Comment (amyeroberts) [inline tests/models/llava/test_image_processing_llava.py:138]: This needs to be more thoroughly tested namely - it needs to test: 

* background colour is properly set when `background_color` is a non-default int, or a non-default tuple
* the method works as expected for both channels first and channels last inputs

File: src/transformers/models/llava/image_processing_llava.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+            "do_normalize",
+            "image_mean",
+            "image_std",
+            "do_convert_rgb",
+            "return_tensors",
+            "data_format",
+            "input_data_format",
+        ]
+
+    def pad_to_square(
```
Comment (amyeroberts) [inline src/transformers/models/llava/image_processing_llava.py:149]: should accept data_format as an input

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

Main comment is about proper testing

File: src/transformers/models/llava/image_processing_llava.py (hunk 7 above)
Comment (NielsRogge) [reply src/transformers/models/llava/image_processing_llava.py:151]: Done

**Comment by hljjjmssyh** [comment]
Is there any update for this feature? It seems very important for aligning implementations between Hugging Face and the official ones.

File: tests/models/llava/test_image_processing_llava.py (hunk 6 above)
Comment (NielsRogge) [reply tests/models/llava/test_image_processing_llava.py:138]: Hi @zucchini-nlp could you take this up?

Comment (zucchini-nlp) [reply tests/models/llava/test_image_processing_llava.py:138]: oke, no prob, I will come back on this tomorrow

File: src/transformers/models/llava/image_processing_llava.py [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+
+        if do_normalize:
+            images = [
+                self.normalize(image=image, mean=image_mean, std=image_std, input_data_format=input_data_format)
+                for image in images
+            ]
+
+        images = [
+            to_channel_dimension_format(image, data_format, input_channel_dim=input_data_format) for image in images
+        ]
```
Comment (qubvel) [inline src/transformers/models/llava/image_processing_llava.py:442]: Lets use a single loop here

File: src/transformers/models/llava/image_processing_llava.py [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+        do_rescale: bool = None,
+        rescale_factor: float = None,
+        do_normalize: bool = None,
+        image_mean: Optional[Union[float, List[float]]] = None,
+        image_std: Optional[Union[float, List[float]]] = None,
+        do_convert_rgb: bool = None,
+        return_tensors: Optional[Union[str, TensorType]] = None,
+        data_format: Optional[ChannelDimension] = ChannelDimension.FIRST,
+        input_data_format: Optional[Union[str, ChannelDimension]] = None,
+        **kwargs,
```
Comment (qubvel) [inline src/transformers/models/llava/image_processing_llava.py:297]:
```suggestion
        do_pad: bool = None,
        do_resize: Optional[bool] = None,
        size: Optional[Dict[str, int]] = None,
        resample: Optional[PILImageResampling] = None,
        do_center_crop: Optional[bool] = None,
        crop_size: Optional[int] = None,
        do_rescale: Optional[bool] = None,
        rescale_factor: Optional[float] = None,
        do_normalize: Optional[bool] = None,
        image_mean: Optional[Union[float, List[float]]] = None,
        image_std: Optional[Union[float, List[float]]] = None,
        do_convert_rgb: Optional[bool] = None,
        return_tensors: Optional[Union[str, TensorType]] = None,
        data_format: Optional[ChannelDimension] = ChannelDimension.FIRST,
        input_data_format: Optional[Union[str, ChannelDimension]] = None,
        **kwargs,
```

File: src/transformers/models/llava/image_processing_llava.py [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+            "image_std",
+            "do_convert_rgb",
+            "return_tensors",
+            "data_format",
+            "input_data_format",
+        ]
+
+    def pad_to_square(
+        self,
+        image: np.array,
```
Comment (qubvel) [inline src/transformers/models/llava/image_processing_llava.py:153]:
```suggestion
        image: np.ndarray,
```

**Review by qubvel — COMMENTED** [review]
Thanks, looks good to me, just a few comments

**Review by qubvel — APPROVED** [review]
